### PR TITLE
Update kubectl rollout proposal to reflect current implementation

### PR DESCRIPTION
**Pull request by janetkuo**

### CODE REVIEW

File: contributors/design-proposals/deploy.md [hunk 1]
```diff
@@ -118,29 +118,29 @@ See issue [#17164](https://github.com/kubernetes/kubernetes/issues/17164).
...
-deployment history, and then we do a rollout with the same PodTemplate as version 1, then version 1 is lost and becomes version 4 after the rollout.
+Note that a Deployment's ReplicaSets use PodTemplate hashes (i.e. the hash of `.spec.template`) to distinguish from each others. When doing rollout or rollback, a Deployment reuses existing ReplicaSet if it has the same PodTemplate, and its `kubernetes.io/change-cause` and `deployment.kubernetes.io/revision` annotations will be updated by the new rollout. At this point, the earlier state of this ReplicaSet is lost in history. For example, if we had 3 ReplicaSets in
+Deployment history, and then we do a rollout with the same PodTemplate as version 1, then version 1 is lost and becomes version 4 after the rollout.
```
Comment (kargakis) [inline contributors/design-proposals/deploy.md]: We have added an annotation that holds the history of the revisions a ReplicaSet has served. Mind referencing it here?
https://github.com/kubernetes/kubernetes/pull/34249

File: contributors/design-proposals/deploy.md [hunk 2]
```diff
@@ -118,29 +118,29 @@ See issue [#17164](https://github.com/kubernetes/kubernetes/issues/17164).
...
 
 ### Pause Deployments
 
-Users sometimes need to temporarily disable a deployment. See issue [#14516](https://github.com/kubernetes/kubernetes/issues/14516).
+Users sometimes need to temporarily disable a Deployment. See issue [#14516](https://github.com/kubernetes/kubernetes/issues/14516).
 
 ### Perm-failed Deployments
 
-The deployment could be marked as "permanently failed" for a given spec hash so that the system won't continue thrashing on a doomed deployment. The users can retry a failed deployment with `kubectl rollout retry`. See issue [#14519](https://github.com/kubernetes/kubernetes/issues/14519).
+The Deployment could be marked as "permanently failed" for a given spec hash so that the system won't continue thrashing on a doomed Deployment. The users can retry a failed Deployment with `kubectl rollout retry`. See issue [#14519](https://github.com/kubernetes/kubernetes/issues/14519).
```
Comment (kargakis) [inline contributors/design-proposals/deploy.md]: I suggest repurposing this section to mention progressDeadlineSeconds and maybe point to our docs about it. Also `kubectl rollout retry` is not a thing so I would remove it from here to avoid any confusion.

File: contributors/design-proposals/deploy.md [hunk 3]
```diff
@@ -118,29 +118,29 @@ See issue [#17164](https://github.com/kubernetes/kubernetes/issues/17164).
...
-- `rollout.kubectl.kubernetes.io/change-source`, which is optional, records the kubectl command of the last mutation made to this rollout. Users may use `--record` in `kubectl` to record current command in this annotation.
-- `rollout.kubectl.kubernetes.io/version` records a version number to distinguish the change sequence of a deployment's
-replication controllers. A deployment obtains the largest version number from its replication controllers and increments the number by 1 upon update or creation of the deployment, and updates the version annotation of its new replication controller.
+We store previous deployment version information in annotations `kubernetes.io/change-cause` and `deployment.kubernetes.io/revision` of ReplicaSets of the Deployment, to support rolling back changes as well as for the users to view previous changes with `kubectl rollout history`.
+- `kubernetes.io/change-cause`, which is optional, records the kubectl command of the last mutation made to this rollout. Users may use `--record` in `kubectl` to record current command in this annotation.
```
Comment (kargakis) [inline contributors/design-proposals/deploy.md]: @soltysh fyi if we ever remove --record this needs to be updated

File: contributors/design-proposals/deploy.md (hunk 1 above)
Comment (janetkuo) [inline contributors/design-proposals/deploy.md]: Done

File: contributors/design-proposals/deploy.md (hunk 2 above)
Comment (janetkuo) [inline contributors/design-proposals/deploy.md]: Done

**Comment by janetkuo** [comment]
@kargakis ptal

File: contributors/design-proposals/deploy.md (hunk 3 above)
Comment (soltysh) [inline contributors/design-proposals/deploy.md]: Roger that. Ref https://github.com/kubernetes/kubernetes/issues/40422.

File: contributors/design-proposals/deploy.md [hunk 4]
```diff
@@ -118,29 +124,35 @@ See issue [#17164](https://github.com/kubernetes/kubernetes/issues/17164).
...
-Users sometimes need to temporarily disable a deployment. See issue [#14516](https://github.com/kubernetes/kubernetes/issues/14516).
+Users sometimes need to temporarily disable a Deployment. See issue [#14516](https://github.com/kubernetes/kubernetes/issues/14516).
+
+For more details, see [pausing and resuming a
+Deployment](https://kubernetes.io/docs/user-guide/deployments/#pausing-and-resuming-a-deployment).
 
 ### Perm-failed Deployments
 
-The deployment could be marked as "permanently failed" for a given spec hash so that the system won't continue thrashing on a doomed deployment. The users can retry a failed deployment with `kubectl rollout retry`. See issue [#14519](https://github.com/kubernetes/kubernetes/issues/14519).
+The Deployment could be marked as "permanently failed" for a given spec hash so that the system won't continue thrashing on a doomed Deployment. See issue [#14519](https://github.com/kubernetes/kubernetes/issues/14519).
```
Comment (kargakis) [inline contributors/design-proposals/deploy.md]: I think we are misusing the term "perma-failed". Once we identify a permanent failure, we should be able to abort the rollout but in reality I don't think we can identify those reliably, that's why we still haven't ended up doing something about it (imagePullBackOff, crashLoopBackOff, insufficient quota, or permissions, all can be transient errors). ProgressDeadlineSeconds merely bubbles a Progressing condition with reason=ProgressDeadlineExceeded in the Deployment status. This can be useful for external processes now and in the future I expect we will automatically rollback as soon as we observe the condition (users now could easily implememnt autoRollback by writting an external controller that would populate d.Spec.Rollback as soon as a ProgressDeadlineExceeded is observed).

TL; DR
Let's stop using "permanently". Switch the title as well to "Failed" or "Stuck".

Comment (janetkuo) [reply contributors/design-proposals/deploy.md]: Updated, PTAL

**Comment by janetkuo** [comment]
Rebased. PTAL @kow3ns @kargakis

**Comment by janetkuo** [comment]
Fixed nits and /s/version/revision/g. @kow3ns PTAL

**Comment by janetkuo** [comment]
@kow3ns need lgtm too

File: contributors/design-proposals/apps/deploy.md [hunk 5]
```diff
@@ -88,53 +94,61 @@ $ kubectl run nginx --image=nginx --replicas=2 --generator=deployment/v1beta1
...
+### Failed Deployments
 
-### Perm-failed Deployments
+The Deployment could be marked as "failed" when it gets stuck trying to deploy
+its newest ReplicaSet without completing within the given deadline (specified
+with `.spec.progressDeadlineSeconds`), see document about
+[failed Deployment](https://kubernetes.io/docs/user-guide/deployments/#failed-deployment).
 
-The deployment could be marked as "permanently failed" for a given spec hash so that the system won't continue thrashing on a doomed deployment. The users can retry a failed deployment with `kubectl rollout retry`. See issue [#14519](https://github.com/kubernetes/kubernetes/issues/14519).
+The users can retry a failed Deployment with `kubectl rollout retry`.
```
Comment (kargakis) [inline contributors/design-proposals/apps/deploy.md]: Is there such a command?
